plugins/database/mongodb/mongodb.go

```diff
@@ -190,6 +203,21 @@ func (m *MongoDB) SetCredentials(ctx context.Context, statements dbplugin.Statem
 	case err == nil:
 	case err == io.EOF, strings.Contains(err.Error(), "EOF"):
 		// Call getConnection to reset and retry query if we get an EOF error on first attempt.
+		session, err := m.getConnection(ctx)
+		if err != nil {
+			return "", "", err
+		}
+		err = session.DB(dialInfo.Database).UpsertUser(&mongoUser)
+		if err != nil {
+			return "", "", err
+		}
+	case strings.Contains(err.Error(), "not master"):
+		// Close connection and reconnect if connected node is not primary.
+		if m.mongoDBConnectionProducer.session != nil {
+			m.mongoDBConnectionProducer.session.Close();
+		}
+		m.mongoDBConnectionProducer.session = nil
```

Comment:
We could move this into the conditional above so that we always nil out the session after we close it.